### PR TITLE
Fixed #34013 -- Added QuerySet.order_by() support for annotation transforms.

### PR DESCRIPTION
**Pull request by charettes**
Thanks Eugene Morozov and Ben Nace for the reports.

### CODE REVIEW

File: django/db/models/sql/compiler.py [hunk 1]
```diff
@@ -387,18 +387,19 @@ def _order_by_pairs(self):
...
-                    expr = F(col)
-                else:
-                    expr = self.query.annotations[col]
-                    if isinstance(expr, Value):
-                        # output_field must be resolved for constants.
-                        expr = Cast(expr, expr.output_field)
+                    expr = F(ref)
+                if isinstance(expr, Value):
+                    # output_field must be resolved for constants.
+                    expr = Cast(expr, expr.output_field)
```
Comment (charettes) [inline django/db/models/sql/compiler.py:402]: A transform could theoretically generate a `Value` expression even if unlikely so I figured I'd add it here.

File: django/db/models/sql/compiler.py [hunk 2]
```diff
@@ -387,18 +387,19 @@ def _order_by_pairs(self):
...
-                # References to an expression which is masked out of the SELECT
-                # clause.
-                if self.query.combinator and self.select:
+
+            ref, *transforms = col.split(LOOKUP_SEP)
+            if expr := self.query.annotations.get(ref):
+                if transforms:
+                    for name in transforms:
+                        expr = self.query.try_transform(expr, name)
+                elif self.query.combinator and self.select:
```
Comment (charettes) [inline django/db/models/sql/compiler.py:396]: Maybe we should error out if trying to order with a transform from a combined query as this is not supported.

**Review by charettes — COMMENTED** [review]
I wonder if we should categorize this as a new feature worth of mention in the release notes instead given it goes beyond `JSONObject`, `ArrayAgg`, and all. It adds generic support for ordering by any annotation transform.

File: tests/aggregation/tests.py [hunk 3]
```diff
@@ -1727,6 +1728,28 @@ def test_aggregation_random_ordering(self):
             ordered=False,
         )
 
+    def test_order_by_aggregate_transform(self):
+        class Mod100(Mod, Transform):
+            def __init__(self, expr):
+                super().__init__(expr, 100)
+
+        sum_field = IntegerField()
+        sum_field.register_instance_lookup(Mod100, "mod100")
```
Comment (charettes) [inline tests/aggregation/tests.py:1737]: Happy to use `django.tests.utils.register_lookup` instead. This felt like a nicer that didn't pollute global state and made use of the relatively recently introduced instance lookups feature.

**Comment by charettes** [comment]
buildbot, test on oracle.

File: django/db/models/sql/compiler.py (hunk 2 above)
Comment (felixxm) [reply django/db/models/sql/compiler.py:396]: Agreed.

**Review by felixxm — COMMENTED** [review]
@charettes Thanks :+1: Can you add a release note?

**Review by felixxm — APPROVED** [review]
@charettes Thanks :+1:

**Pull request merged**